src/codecs/avif/decoder.rs

```diff
@@ -102,93 +324,231 @@ impl<R: Read> ImageDecoder for AvifDecoder<R> {
+                } else {
+                    yuv400_to_rgba12
+                }
+            }
+            PixelLayout::I420 => {
+                if bit_depth == 10 {
+                    yuv420_to_rgba10
+                } else {
+                    yuv420_to_rgba12
+                }
+            }
+            PixelLayout::I422 => {
+                if bit_depth == 10 {
+                    yuv422_to_rgba10
+                } else {
+                    yuv422_to_rgba12
+                }
+            }
+            PixelLayout::I444 => {
+                if bit_depth == 10 {
+                    yuv444_to_rgba10
+                } else {
+                    yuv444_to_rgba12
+                }
+            }
+        };
+        worker(image, target, yuv_range, color_matrix)?;
+
+        // Squashing alpha plane into a picture
+        if let Some(picture) = &self.alpha_picture {
             if picture.pixel_layout() != PixelLayout::I400 {
-                return Err(ImageError::Unsupported(
-                    UnsupportedError::from_format_and_kind(
-                        ImageFormat::Avif.into(),
-                        UnsupportedErrorKind::GenericFeature(format!(
-                            "Alpha must be PixelLayout::I400 but was: {:?}",
-                            picture.pixel_layout() // PixelLayout does not implement display
-                        )),
-                    ),
-                ));
+                return Err(ImageError::Decoding(DecodingError::new(
+                    ImageFormat::Avif.into(),
+                    AvifDecoderError::AlphaPlaneFormat(picture.pixel_layout()),
+                )));
             }
-            let stride = picture.stride(PlanarImageComponent::Y) as usize;
-            let plane = picture.plane(PlanarImageComponent::Y);
-            let width = picture.width();
+
+            let a_dav1d_plane = picture.plane(PlanarImageComponent::Y);
+            let a_plane_view = transmute_y_plane16(
+                &a_dav1d_plane,
+                picture.stride(PlanarImageComponent::Y) as usize,
+                width as usize,
+                height as usize,
+            );
+
             for (buf, slice) in Iterator::zip(
-                buf.chunks_exact_mut(width as usize * 4),
-                plane.as_ref().chunks_exact(stride),
+                target.chunks_exact_mut(width as usize * 4),
+                a_plane_view.data.as_ref().chunks_exact(a_plane_view.stride),
             ) {
-                for i in 0..width as usize {
-                    buf[3 + i * 4] = slice[i];
+                for (rgba, a_src) in buf.chunks_exact_mut(4).zip(slice) {
+                    rgba[3] = *a_src;
                 }
             }
         }
 
-        Ok(())
-    }
+        // Expand current bit depth to target 16
+        let target_expand_bits = 16u32 - self.picture.bit_depth() as u32;
+        for item in target.iter_mut() {
+            *item <<= target_expand_bits;
```

Comment:
When increasing the bit depth, I think we actually want to replicate the high bits: 0x000 -> 0x0000, 0x123 -> 0x1231, 0xfff -> 0xffff.

Should be something like this:
```rust
*item = (*item << target_expand_bits) | (*item >> (16 - target_expand_bits));
```

Reply:
Thanks for suggestion!